src/embedded_flash/mod.rs

```diff
@@ -12,31 +12,41 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#[cfg(feature = "std")]
+mod buffer_upgrade;
+mod helper;
 #[cfg(not(feature = "std"))]
 mod syscall;
+mod upgrade_storage;
 
-#[cfg(not(feature = "std"))]
-pub use self::syscall::SyscallStorage;
+pub use upgrade_storage::UpgradeStorage;
 
-/// Storage definition for production.
+/// Definitions for production.
 #[cfg(not(feature = "std"))]
 mod prod {
-    pub type Storage = super::SyscallStorage;
+    use super::syscall::{SyscallStorage, SyscallUpgradeStorage};
+
+    pub type Storage = SyscallStorage;
 
     pub fn new_storage(num_pages: usize) -> Storage {
         Storage::new(num_pages).unwrap()
     }
+
+    pub type UpgradeLocations = SyscallUpgradeStorage;
 }
 #[cfg(not(feature = "std"))]
-pub use self::prod::{new_storage, Storage};
+pub use self::prod::{new_storage, Storage, UpgradeLocations};
 
-/// Storage definition for testing.
+/// Definitions for testing.
 #[cfg(feature = "std")]
 mod test {
+    use super::buffer_upgrade::BufferUpgradeStorage;
+
+    const PAGE_SIZE: usize = 0x1000;
```

Comment:
Let's put this back inside `new_storage` if it's not used? I don't see any reason it should be shared.

Reply:
True, I forgot. Done.